### ISSUE TITLE
Deleting a Namespace of a project with a Shoot causes deadlock

### ISSUE THREAD

**Issue by bergerx** [gardener/gardener#237]
I accidentally deleted the Namespace of a Project with a Shoot inside it. After some time Namespace stuck in Terminating.

TL;DR: namespace blocked on shoot, shoot blocked on bakcup, backup blocked on namespace termination

Here is a related message from gardener-controller-manager:
> Description:  Failed to delete Shoot cluster: Errors occurred during flow execution: '(*Botanist).DeploySecrets' returned 'secrets "man.kubeconfig" is forbidden: unable to create new content in namespace man because it is being terminated.', '(*Botanist).DeployBackupInfrastructure' returned 'backupinfrastructures.garden.sapcloud.io "shoot--man--man--d18ac" is forbidden: unable to create new content in namespace man because it is being terminated'

Here is how the deadlock happened:
- the namespace is marked for deletion and no resources is allowed to be created in that namespace,
- namespace fired delete on all resources inside it including the shoot, and namespace is waiting on termination until everything inside is removed,
- shoot removal triggered a BackupInfrastructure resource generation which is not possible because the namespace is marked for deletion

I found myself with a hard-to-recover problem in my hand. Once the namespace is marked for deletion there is no way back, which means the shoot has to be deleted but can't.

This issue should be explicitly handled by `gardener-controller-manager`.

**Comment by bergerx** [comment]
Some options I see:
- prevent the premature Shoot deletion (being deleted without annotation) until operators issue a Shoot recovery or confirm the deletion, which could be challenging because a new backup cant be issued.
- take the shoot deletion as granted without a backup generation triggered, let everything to be deleted gracefully.

**Comment by bergerx** [comment]
We just had this issue in another cluster as a result of an operator mistake, last time when we hist this issue we solved it by destroying the whole gardener setup, but this time we have some other shoots on the gerdener that we'd prefer to keep, any idea about how to unblock the ns termination.

Here is the current status:
```
[msa-infra-garden:msa-infra]k8s-clusters:msa-infra (master $ u=) $ kubectl get ns
NAME           STATUS        AGE
azure-system   Active        62d
default        Active        62d
garden         Active        62d
kube-public    Active        62d
kube-system    Active        62d
msa-infra      Terminating   62d
msa-workshop   Active        5d
[msa-infra-garden:msa-infra]k8s-clusters:msa-infra (master u=) $ kubectl get --all-namespaces shoots | sed 's/dub.*io/<cropped>/'
NAMESPACE   NAME          CLOUDPROFILE   VERSION   SEED              DOMAIN                  OPERATION   PROGRESS   CONTROL   NODES     SYSTEM    AGE
msa-infra   dex-infra     azure          1.11.0    azure-eastus-01   dex-infra.<cropped>     Failed      28         True      True      False     27d
msa-infra   infra-test    azure          1.11.0    azure-eastus-01   infra-test.<cropped>    Failed      28         True      True      True      62d
msa-infra   infra-test2   azure          1.10.5    azure-eastus-01   infra-test2.<cropped>   Failed      28         True      False     False     14d
msa-infra   one-dex-new   azure          1.11.0    azure-eastus-01   one-dex.<cropped>       Failed      28         True      True      True      15d
msa-infra   two-dex       azure          1.11.0    azure-eastus-01   two-dex.<cropped>       Failed      28         True      False     False     19d
[msa-infra-garden:msa-infra]k8s-clusters:msa-infra (master $ u=) $ 
```

Reconciliations for all shoots in that namespace are stuck on DeployBackupInfrastructure flow step, since it's trying to create a secret:
```
[msa-infra-garden:msa-infra]k8s-clusters:msa-infra (master $ u=) $ gardenctl ls issues
issues:
- project: msa-infra
  seed: azure-eastus-01
  shoot: dex-infra
  health: NotReady
  status:
    lastError: 'Failed to reconcile Shoot cluster state: Errors occurred during flow
      execution: ''(*Botanist).DeployBackupInfrastructure'' returned ''backupinfrastructures.garden.sapcloud.io
      "shoot--msa-infra--dex-infra--4852b" is forbidden: unable to create new content
      in namespace msa-infra because it is being terminated'', ''(*Botanist).DeploySecrets''
      returned ''secrets "dex-infra.kubeconfig" is forbidden: unable to create new
      content in namespace msa-infra because it is being terminated'''
    lastOperation:
      description: 'Failed to reconcile Shoot cluster state: Errors occurred during
        flow execution: ''(*Botanist).DeployBackupInfrastructure'' returned ''backupinfrastructures.garden.sapcloud.io
        "shoot--msa-infra--dex-infra--4852b" is forbidden: unable to create new content
        in namespace msa-infra because it is being terminated'', ''(*Botanist).DeploySecrets''
        returned ''secrets "dex-infra.kubeconfig" is forbidden: unable to create new
        content in namespace msa-infra because it is being terminated'''
      lastUpdateTime: 2018-08-24 18:45:32 +0100 IST
      progress: 28
      state: Failed
      type: Reconcile
...
```

Any suggestion to unblock the status and let the shoots deleted?

**Comment by rfranzke** [comment]
Save the YAMLs of the `Shoot`s and `SecretBinding`s in the `msa-infra` namespace into a file. Then delete all `Shoot`s and `BackupInfrastructure`s manually, after that recreate the namespace:

```bash
$ kubectl -n msa-infra get shoot -o yaml > /tmp/msa-infra-shoots
$ kubectl -n msa-infra get sb -o yaml > /tmp/msa-infra-secretbindings
$ kubectl -n msa-infra delete shoot --all
$ kubectl -n msa-infra edit shoot # manually remove the finalizer from all shoots
$ kubectl -n msa-infra delete backupinfra --all
$ kubectl -n msa-infra edit backupinfra # manually remove the finalizer from all backupinfras
# wait until the namespace has been deleted, then recreate the namespace
$ kubectl create ns msa-infra
# create your azure cloud provider secret here
$ kubectl apply -f /tmp/msa-infra-secretbindings
$ kubctl apply -f /tmp/msa-infra-shoots
```

Gardener should pick up the existing clusters and reconcile them. The only thing that is gone and will be recreated is the ABS containers holding the etcd backups.
PS: You have to manually delete the old ABS containers from the Azure console.

**Comment by bergerx** [comment]
Thanks a lot @rfranzke , i didn't follow your instructions exactly but i got the overall idea and that helped me to solve the issue.

Here is more or less what i did:
* suspended gardener-apiserver and gardener-controller-manager (scaled both to 0), verified the pods are terminated
* deleted shoot and secret bindings from gardener's own etcd (we have a dedicated etcd for gardener) using commands like `kubectl exec -it  etcd-msa-infra-0 -c etcd -- sh -c 'ETCDCTL_API=3 etcdctl del /registry/garden.sapcloud.io/garden.sapcloud.io/shoots/msa-infra/dex-infra'`
* scaled gardener-apiserver back up (i left gardener-controller-manager suspended at this stage) and expecting no shoots reported, but shoot queries were still returning the shoots even though gardener's etcd has no records, found it incredibly strange and re-checked the gardener's etcd, again no shoot records, seems like kubernetes itself has a kind of cache on its side as well
* i manually removed the gardener related finalizers from all resources in the namespace
* issued a `kubectl delete` but they failed with `forbidden: must have a "confirmation.garden.sapcloud.io/deletion" annotation to delete`
* annotated all resources with `kubectl annotate all --all -n msa-infra confirmation.garden.sapcloud.io/deletion=true`
* re-issued `kubectl delete` on all resources, they are all removed immediately
* namespace termination went through
* manually deleted all shoot resource groups on Azure portal (thanks to the Azure's resource group concept)
* resumed gardener-controller-manager (scaled to 1)
